### PR TITLE
Fixes of bugs found during February 2023 conference/workshops

### LINKED ISSUE #3764: Zoom to extent doesn't work for geojsons loaded from file
# Bug

**Describe the bug**
Zoom to extent doesn't work for geojsons loaded from file

EDIT: geoJSONs uploaded to Layman and then synced

**Expected behavior**
Should zoom to extent

### PR DESCRIPTION
**Pull request by FilipLeitner**
fixes #3764
## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Dependency updates
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)
- [ ] I am unsure (we'll look into it together)

## Do you introduce a breaking change?

- [ ] Yes
- [x] No
- [ ] I am unsure (no worries, we'll find out)

## Checklist

- [ ] I understand and agree that the changes in this PR will be licensed under the [MIT License]
- [ ] I have followed the [guidelines for contributing](https://github.com/hslayers/hslayers-ng/blob/master/CONTRIBUTING.md)
- [ ] The proposed change fits to the content of the [code of conduct](https://github.com/hslayers/hslayers-ng/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have added or updated tests and documentation, and the test suite passes (run `npm test` locally)
- [ ] I'm lost; why do I have to check so many boxes? Please help!

### CODE REVIEW

File: projects/hslayers/src/components/query/query-wmts.service.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ export class HsQueryWmtsService {
   constructor(
     private hsMapService: HsMapService,
     private hsUtilsService: HsUtilsService
-  ) {}
+  ) { }
```
Comment (jmacura) [inline projects/hslayers/src/components/query/query-wmts.service.ts:19]: lint

File: projects/hslayers/src/components/query/query-wmts.service.ts [hunk 2]
```diff
@@ -29,7 +29,7 @@ export class HsQueryWmtsService {
     layer: Layer<WMTS>,
     coordinate: number[],
     app: string
-  ): Promise<{url: string; format: string}> {
+  ): Promise<{ url: string; format: string }> {
```
Comment (jmacura) [inline projects/hslayers/src/components/query/query-wmts.service.ts:32]: lint

File: projects/hslayers/src/components/save-map/interfaces/layman-layer-descriptor.interface.ts [hunk 3]
```diff
@@ -86,4 +86,6 @@ export interface HsLaymanLayerDescriptor {
    * Only when error
    */
   sub_code?: number;
+  //Not part of origial response. Added by describeLayer
```
Comment (jmacura) [inline projects/hslayers/src/components/save-map/interfaces/layman-layer-descriptor.interface.ts:89]:
```suggestion
  /**
   * Not part of origial response. Added by describeLayer
   */
```

**Review by jmacura — CHANGES_REQUESTED** [review]
Default imports are not tree-shakeable and were causing some issues with federated modules (see https://github.com/hslayers/hslayers-ng/pull/3514)

I have already done the change, no action needed from you @FilipLeitner in this matter.

**Review by jmacura — APPROVED** [review]

**Comment by jmacura** [comment]
Code-wise the change is fine with me. Thanks again for all the fixes!

I have tested the functionality and there is one missing imperfection regarding #3764 related changes. When I add WFS layer from an external source (via Add Data -> URL), the layer is not zoomed in, while when I add the same layer from Data Catalogue, it does zoom correctly.

**Comment by jmacura** [comment]
> (...) When I add WFS layer from an external source (via Add Data -> URL), the layer is not zoomed in, while when I add the same layer from Data Catalogue, it does zoom correctly.

Ah, hint 😄 https://github.com/hslayers/hslayers-ng/blob/ac50d6156d5a2b94fd830f67dc055e0cc275e6d7/projects/hslayers/src/components/add-data/common/url/add/add.component.ts#L53

**Comment by FilipLeitner** [comment]
Oh sure, I tunnelled hard just on the file upload. I will fix this as well :)

**Comment by FilipLeitner** [comment]
Took much longer than expected. WFS and WMS seem to be working fine.  Rest of the layer types tbd

File: projects/hslayers/src/components/add-data/url/add-data-url.service.ts [hunk 4]
```diff
@@ -25,7 +27,8 @@ export class HsAddDataUrlService {
   constructor(
     public hsLog: HsLogService,
     public hsLanguageService: HsLanguageService,
-    public hsLayoutService: HsLayoutService
+    public hsLayoutService: HsLayoutService,
+    private hsMapsService: HsMapService
```
Comment (jmacura) [inline projects/hslayers/src/components/add-data/url/add-data-url.service.ts:31]:
```suggestion
    private hsMapService: HsMapService
```
With code-completion, this is not that much of a issue, but could be misleading..

File: projects/hslayers/src/components/add-data/url/wfs/wfs.service.ts [hunk 5]
```diff
@@ -211,12 +217,46 @@ export class HsUrlWfsService implements HsUrlTypeServiceModel {
...
+      return selectedLayerNames.includes(lyr.Name);
+    });
+    const layerExtents: number[][] = layers.map((lyr) => {
+      return this.getLayerExtent(lyr, appRef.data.map_projection, app);
+    });
+    return this.hsAddDataUrlService.calcCombinedExtent(layerExtents);
+  }
+
+  /**
+   * Zoom map to one layers or combined layer list extent
```
Comment (jmacura) [inline projects/hslayers/src/components/add-data/url/wfs/wfs.service.ts:252]:
```suggestion
   * Zoom map to one layer or combined layer list extent
```

File: projects/hslayers/src/components/add-data/url/wms/wms.service.ts [hunk 6]
```diff
@@ -290,60 +288,45 @@ export class HsUrlWmsService implements HsUrlTypeServiceModel {
...
-          serviceLayer.BoundingBox[0].crs || crs, //Use BBOX object crs - when missing assume its same as layer's
-          this.hsMapService.getMap(app).getView().getProjection()
+
+      if (preferred?.extent) {
+        boundingbox = preferred?.extent;
+      } else {
+        boundingbox = boundingbox.filter((b) => b.crs != 'CRS:84');
+        boundingbox = transformExtent(
+          boundingbox[0].extent,
+          boundingbox[0].crs || crs, //Use BBOX object crs - when missing assume its same as layer's
```
Comment (jmacura) [inline projects/hslayers/src/components/add-data/url/wms/wms.service.ts:320]:
```suggestion
          boundingbox[0].crs || crs, //Use BBOX object crs - when missing assume it's same as layer's
```

File: projects/hslayers/src/components/query/query-wms.service.ts [hunk 7]
```diff
@@ -232,113 +192,84 @@ export class HsQueryWmsService {
...
-            }
-          }
-          this.updateFeatureList(updated, group, app);
-        }
-      }
-    }
+    let updated = false;
+    features.forEach((feature) => {
+      /**
+       * TODO: Layered response need to be refactored as well but I havent found and example yet
```
Comment (jmacura) [inline projects/hslayers/src/components/query/query-wms.service.ts:234]:
```suggestion
       * TODO: Layered response need to be refactored as well but I haven't found an example yet
```

**Review by jmacura — COMMENTED** [review]
Some minor typos again, but good to go from my POV.

Thanks for the improvement

**Pull request merged**